sharkfuser/include/fusilli/node/matmul_node.h

```diff
@@ -0,0 +1,182 @@
+
+class MatmulNode : public NodeCRTP<MatmulNode> {
+public:
+  MatmulAttr matmulAttr;
+
+  MatmulNode(MatmulAttr &&attr, const Context &ctx)
+      : NodeCRTP(ctx), matmulAttr(std::move(attr)) {}
+
+  const std::string &getName() const override final {
+    return matmulAttr.getName();
+  }
+  Type getType() const override final { return Type::Matmul; }
+
+  ErrorObject preValidateNode() const override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Pre-Validating MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    // Ensure input and output tensors are set.
+    FUSILLI_RETURN_ERROR_IF(!aT, ErrorCode::AttributeNotSet,
+                            "Matmul input tensor A not set");
+    FUSILLI_RETURN_ERROR_IF(!bT, ErrorCode::AttributeNotSet,
+                            "Matmul input tensor B not set");
+    FUSILLI_RETURN_ERROR_IF(!cT, ErrorCode::AttributeNotSet,
+                            "Matmul output tensor C not set");
+
+    size_t aRank = aT->getDim().size();
+    size_t bRank = bT->getDim().size();
+
+    // Rank checks on input tensors (must be at least rank 2).
+    FUSILLI_RETURN_ERROR_IF(
+        aRank < 2, ErrorCode::InvalidAttribute,
+        "Matmul input tensor A must have a rank of at least 2");
+    FUSILLI_RETURN_ERROR_IF(
+        bRank < 2, ErrorCode::InvalidAttribute,
+        "Matmul input tensor B must have a rank of at least 2");
+
+    // Check that inner dimensions match (K dimension).
+    const std::vector<int64_t> &aDim = aT->getDim();
+    const std::vector<int64_t> &bDim = bT->getDim();
+    int64_t aK = aDim[aRank - 1]; // Last dimension of A
+    int64_t bK = bDim[bRank - 2]; // Second-to-last dimension of B
+
+    FUSILLI_RETURN_ERROR_IF(
+        aK != bK, ErrorCode::InvalidAttribute,
+        "Matmul input tensors A and B have incompatible inner dimensions (K): "
+        "A has K=" +
+            std::to_string(aK) + ", B has K=" + std::to_string(bK));
+
+    return ok();
+  }
+
+  ErrorObject inferPropertiesNode() override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Inferring properties for MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    matmulAttr.fillFromContext(context);
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    const std::vector<int64_t> &aDim = aT->getDim();
+    const std::vector<int64_t> &bDim = bT->getDim();
+
+    const std::vector<int64_t> &cDim = cT->getDim();
+    const std::vector<int64_t> &cStride = cT->getStride();
+
+    // Infer shape of output tensor.
+    if (cDim.empty())
+      cT->setDim(getMatmulInferredOutputShape(aDim, bDim));
+
+    // Infer stride of output tensor.
```

Comment:
nit: Consider revising the comment to reflect that a contiguous / row-major stride is picked when un-specified.